dxr/query.py

```diff
@@ -38,19 +69,24 @@ def __init__(self, es_search, querystr, enabled_plugins):
 
         # A list of dicts describing query terms:
         grammar = query_grammar(self.enabled_plugins)
-        self.terms = QueryVisitor().visit(grammar.parse(querystr))
+        self._terms = QueryVisitor().visit(grammar.parse(querystr))
 
-    def single_term(self):
+        enabled_filters_by_name = filters_by_name(self.enabled_plugins)
+        # A list of lists, each inner list representing the filters of the name
+        # of the parallel term. We OR the elements of the inner lists and then
+        # AND those OR balls together.
+        self._filters = [[f(term, self.enabled_plugins) for f in
+                          enabled_filters_by_name[term['name']]] for term in self._terms]
+
+    def single_text_term(self):
         """Return the single, non-negated textual term in the query.
 
-        If there is more than one term in the query or if the single term is a
-        non-textual one, return None.
+        If there is not exactly one such term, return None.
 
         """
-        if len(self.terms) == 1:
-            term = self.terms[0]
-            if term['name'] == 'text' and not term['not']:
-                return term
+        text_terms = [term for term in self._terms if term['name'] == 'text' and not term['not']]
```

Comment:
For example it's common to exclude the object directory in a search, and it can be helpful to still have the promoted results.